File: Lambert.cpp
/*
 *  Lambert.cpp
 *  RayTracing
 *
 *  Created by Camille on 23/01/08.
 *  Copyright 2008 __MyCompanyName__. All rights reserved.
 *
 */

#include "Lambert.h"


Color Lambert::getColor(Point* point, Vector* norm, Ray* ray, std::list<Light* > lights) const{
	
	Color color(od);
	
	for(std::list<Light* >::iterator iter = lights.begin(); iter != lights.end(); ++iter){
		Vector L = getL(point, (*iter));
		
		Color diffuse = ((*(*iter)->getColor()) * (kd * (L * norm)));
		
		color = color * diffuse;
	}
	
	color.normalize();
	
	return color;
}

File: Java/puls4r/enlightment/direct/Lambert.java
/**
 * 
 */
package puls4r.enlightment.direct;

import java.util.Vector;

import javax.vecmath.Color3f;
import javax.vecmath.Point3d;
import javax.vecmath.Vector3d;

import puls4r.enlightment.Enlightment;
import puls4r.scene.objects.Light;
import puls4r.tracer.Ray;

/**
 * @author cmaurice2
 *
 */
public class Lambert extends Enlightment {

	Color3f od;
	double kd;
	
	public Lambert(Color3f od, double kd) {
		this.od = od;
		this.kd = kd;
	}

	@Override
	public Color3f getColor(Point3d point, Vector3d norm, Ray ray,
			Vector<Light> lights) {
		
		Color3f color = new Color3f();
		
		for(Light light : lights){
			Vector3d L = getL(point, light);
			
			Color3f diffuse = new Color3f(light.getColor());
			diffuse.scale((float) (kd * L.dot(norm)) );
			
			color.set(od.x * diffuse.x, od.y * diffuse.y, od.z * diffuse.z);
			Enlightment.normalize(color);
		}
		
		return color;
	}

}
